drivers/timer/cortex_m_systick.c

```diff
@@ -41,6 +41,7 @@ static u32_t elapsed(void)
 	do {
 		val = SysTick->VAL & COUNTER_MAX;
 		ctrl_cache |= SysTick->CTRL;
+		__ISB();
```

Comment:
From: http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0553a/CHDEBIEG.html
*ISB acts as an instruction synchronization barrier. It flushes the pipeline of the processor, so that all instructions following the ISB are fetched from cache or memory again, after the ISB instruction has been completed.*

ISB does not guarantee that the interrupt will be executed before "SysTick->VAL" access (which I think is purpose of this change). IMHO you will see the same effect if you insert some NOPs here.

Reply:
To be more general: We should not rely on the fact that given interrupt is executed at specific point in time without enforcing such thing using standard synchronization mechanisms.

Reply:
@pizi-nordic Please refer to the the following link 
[https://static.docs.arm.com/dai0321/a/DAI0321A_programming_guide_memory_barriers_for_m_profile.pdf](https://static.docs.arm.com/dai0321/a/DAI0321A_programming_guide_memory_barriers_for_m_profile.pdf)  page 10 and page 25.

Page 10 states that "In addition to the ISB instruction, the architecture defines exception entry and exception return to also have ISB semantics, causing a fresh fetch of instructions and a **re-evaluation of interrupts**."
Which means that when a ISB is executed it will check if there are any pending interrupts and it will get serviced according to the priority.

Page 25 states that "if  a pended interrupt request needs to be recognized immediately after being enabled in the NVIC, add a DSB instruction and then an ISB instruction."
Since we are not enabling the interrupt via software we don't need a DSB but we do need a ISB to ensure we get the interrupt when it is ready. 
This is a recommendation by ARM and we shouldn't assume that all ARM soc vendors have this recommendation in place. 


Just a FYI. the sanitycheck is failing on the mainline for qemu because of this bug.